contracts/src/v0.8/dev/automation/2_1/KeeperRegistry2_1.sol

```diff
@@ -461,42 +449,54 @@ contract KeeperRegistry2_1 is KeeperRegistryBase2_1, OCR2Abstract, Chainable, ER
    */
   function _validateBlockTrigger(
     uint256 upkeepId,
-    BlockTrigger memory trigger,
+    bytes memory rawTrigger,
     Upkeep memory upkeep
   ) internal returns (bool) {
-    if (trigger.blockNum < upkeep.lastPerformed) {
+    BlockTrigger memory trigger = abi.decode(rawTrigger, (BlockTrigger));
+    if (trigger.blockNum < upkeep.lastPerformedBlockNumberOrTimestamp) {
       // Can happen when another report performed this upkeep after this report was generated
-      emit StaleUpkeepReport(upkeepId);
+      emit StaleUpkeepReport(upkeepId, rawTrigger);
       return false;
     }
     if (_blockHash(trigger.blockNum) != trigger.blockHash) {
       // Can happen when the block on which report was generated got reorged
       // We will also revert if checkBlockNumber is older than 256 blocks. In this case we rely on a new transmission
       // with the latest checkBlockNumber
-      emit ReorgedUpkeepReport(upkeepId);
+      emit ReorgedUpkeepReport(upkeepId, rawTrigger);
       return false;
     }
     return true;
   }
 
-  function _validateLogTrigger(uint256 upkeepId, LogTrigger memory trigger) internal returns (bool) {
+  function _validateLogTrigger(uint256 upkeepId, bytes memory rawTrigger) internal returns (bool) {
+    LogTrigger memory trigger = abi.decode(rawTrigger, (LogTrigger));
     if (_blockHash(trigger.blockNum) != trigger.blockHash) {
-      emit ReorgedUpkeepReport(upkeepId);
+      emit ReorgedUpkeepReport(upkeepId, rawTrigger);
+      return false;
+    }
+    bytes32 logID = keccak256(abi.encodePacked(trigger.txHash, trigger.logIndex));
+    if (s_observedLogTriggers[logID]) {
+      emit StaleUpkeepReport(upkeepId, rawTrigger);
       return false;
     }
-    // TODO - dedup log processing
+    s_observedLogTriggers[logID] = true;
     return true;
   }
 
-  function _validateCronTrigger(uint256 upkeepId, uint256 trigger, Upkeep memory upkeep) internal returns (bool) {
-    if (trigger < upkeep.lastPerformed) {
+  function _validateCronTrigger(
+    uint256 upkeepId,
+    bytes memory rawTrigger,
+    Upkeep memory upkeep
+  ) internal returns (bool) {
+    uint256 trigger = abi.decode(rawTrigger, (uint256));
+    if (trigger <= upkeep.lastPerformedBlockNumberOrTimestamp) {
       // Can happen when another report performed this upkeep after this report was generated
-      emit StaleUpkeepReport(upkeepId);
+      emit StaleUpkeepReport(upkeepId, rawTrigger);
       return false;
     }
     if (trigger > block.timestamp) {
       // Rare condition where reorged block can have timestamp < than triggering block
-      emit ReorgedUpkeepReport(upkeepId);
+      emit ReorgedUpkeepReport(upkeepId, rawTrigger);
       return false;
```

Comment:
not fully sure if this makes sense. Can discuss cron validation in a separate PR in detail